README.md
# PSQueue (Promises Sync Queue)

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/cgadam/psqueue?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
[![dependencies](https://david-dm.org/cgadam/psqueue.png)](https://david-dm.org/cgadam/psqueue)

### Getting async ducks in a row.

Have you ever find yourself dealing with this problem: you need to execute async operations (encapsulated in Q promises)
in order. The bad thing is that those async operations are dynamically generated and therefore it's not trivial
to chain Q promises with then, then, then.

If this is your case, then this module is for you.

## How to use it?

1. Create a PSQueue.
2. Create your function that returns a Q promise.
3. Add the function to the queue (indicating the parameters)
4. Repeat steps 2 and 3 as many times as you want.
5. Call PSQueue's method resolveAllSync() which will return a Q promise too.

The promise returned by resolveAllSync() will be resolved once ALL the Q promises in the
queue are solved. The good part is that the sync queue resolves all the promises in ORDER: ONE AT THE TIME! :)

The resolved value the promise returned by resolveAllSync() will contain an array with the resolved values of
all the other synchronously executed promises of the queue.

You can optionally define a *torch* object and pass it to resolveAllSync function. The object will be systematically passed between the functions in the queue.
Functions can then modify the value and pass it to the following functions in the queue. You need to be sure that your function will actually proccess this parameter.

Example:

**Q.all** resolves all the promises *asynchronously*. So, in this example:

```javascript
    var Q = require('q');

    var timestamp = (new Date()).getTime();
    var arr = ['A', 'B', 'C'];

    function doPromiseToItem(item) {
      var deferred = Q.defer();
      var randomFactor = 10000 * Math.random()
      setTimeout(function (rand) {
        console.log(item + ' solved after random ' + (Math.round((rand / 1000) * 100) / 100) + ' secs.');
        deferred.resolve(item);
      }, randomFactor, randomFactor);
      return deferred.promise;
    }

    var promises = [];

    for (var i = 0; i < arr.length; i++) {
      var item = arr[i];
      var promise = doPromiseToItem(item);
      promises.push(promise);
    }

    Q.all(promises).then(function (arrResults) {
      var duration = Math.round(((((new Date()).getTime()) - timestamp) / 1000) * 100) / 100;
      console.log('ALL SOLVED in ' + duration + ' seconds.');
      console.log('Results are:\n' + arrResults.join('\n'));
    });
```
If you execute it:

```bash
    $ cd demos
    $ node QAllTest.js
```
This is one possible output you might get:

```bash
    C solved after random 3.03 secs.
    B solved after random 5.09 secs.
    A solved after random 5.45 secs.
    ALL SOLVED in 5.46 seconds.
    Results are:
    A
    B
    C
```

As you can see *all of the promises ran in parallel* (not in a specific order -> C,B,A) and the time it took to complete
all of them is nothing but the time it took for the latest one to finish.

**PSQueue** runs all the promises *in order* and *one at the time*:
(Example using a torch value)

```javascript
    var Q = require('q'),
      PSQueue = require('../psqueue'),
      timestamp = (new Date()).getTime(),
      arr = ['A', 'B', 'C'],
      promisesQueue = new PSQueue();

    function doPromiseToItem(item, torch) {
      var deferred = Q.defer();
      var randomFactor = 10000 * Math.random();
      setTimeout(function (rand) {
        console.log(item + ' solved after ' + torch.text + ' taking random ' + (Math.round((rand / 1000) * 100) / 100) + ' secs.');
        torch.text = (torch.text === 'BEGGINING OF TIME' ? item : (torch.text + ',' + item));
        deferred.resolve(item + ':OK');
      }, randomFactor, randomFactor);
      return deferred.promise;
    }

    for (var i = 0; i < arr.length; i++) {
      var item = arr[i];
      promisesQueue.addPromise(doPromiseToItem, item);
    }

    promisesQueue.resolveAllSync({
      text: 'BEGGINING OF TIME'
    }).then(function (arrResults) {
      var duration = Math.round(((((new Date()).getTime()) - timestamp) / 1000) * 100) / 100;
      console.log('ALL SOLVED in ' + duration + ' seconds.');
      console.log('Results are:\n' + arrResults.join('\n'));
    });
```
If you execute it:

```bash
    $ cd demos
    $ node PSQueueTest.js
```
This is one possible output you might get:

```bash
    A solved after BEGGINING OF TIME taking random 9.6 secs.
    B solved after A taking random 5.55 secs.
    C solved after A,B taking random 1.1 secs.
    ALL SOLVED in 16.3 seconds.
    Results are:
    A:OK
    B:OK
    C:OK
```

As you can see no matter how much time each individual promise took to be solved, all of them are executed
in order (A,B,C) and the total time is the sum of the time of each of them. Also you can see how the torch value is modified from function to function appening
the name of the item that was previosuly executed. The result of the promises is then returned in an array (arrResults).

**What if a promise in the queue fails?**

If one promise in the queue fails, then the following promises in the queue are not executed and the rejection value is used for rejecting the queue's promise.

Example:

```javascript
    var Q = require('q'),
      PSQueue = require('../psqueue'),
      timestamp = (new Date()).getTime(),
      arr = ['A', 'B', 'C', 'D'],
      promisesQueue = new PSQueue();

    function doPromiseToItem(item, torch) {
      var deferred = Q.defer();
      var randomFactor = 10000 * Math.random();
      setTimeout(function (obj) {
        if (obj.item === 'C') {
          console.log(item + ' NOT solved after random ' + (Math.round((obj.rand / 1000) * 100) / 100) + ' secs.');
          deferred.reject(obj.item);
          return;
        }
        console.log(item + ' solved after random ' + (Math.round((obj.rand / 1000) * 100) / 100) + ' secs.');
        torch.push(item);
        deferred.resolve();
      }, randomFactor, {
        "rand": randomFactor,
        "item": item
      });
      return deferred.promise;
    }

    for (var i = 0; i < arr.length; i++) {
      var item = arr[i];
      promisesQueue.addPromise(doPromiseToItem, item);
    }

    var partialResult = [];

    promisesQueue.resolveAllSync(partialResult).fail(function (err, arrResults) {
      var duration = Math.round(((((new Date()).getTime()) - timestamp) / 1000) * 100) / 100;
      console.log('FAILED because I found ' + err + ' after ' + duration + ' seconds.');
      console.log('Partial result is: ' + partialResult.join(','));
    });
```

If you execute it:

```bash
    $ cd demos
    $ node PSQueueTestFail.js
```
This is one possible output you might get:

```bash
    A solved after random 7.35 secs.
    B solved after random 5.06 secs.
    C NOT solved after random 4.76 secs.
    FAILED because I found C after 17.2 seconds.
    Partial result is: A,B
```

As you can see **D** was never executed and the torch object has been used to obtain a possible partialResult.

### Alias for resolveAllSync
You can use *all* and *allSync* as alias of resolveAllSync


CHANGELOG
v0.0.6:
  date: 2015-10-07
  changes:
    - It is possible now to get the list of result values.
    - Introducing "torch" objects that can be passed among functions in the queue.
    - Possible exceptions of the functions in the queue are now handled and returned as part of rejection value.
    - Improved documentation adding dependencies state.
    - Introducing alias all and allSync for resolveAllSync


test/PSQTest.js
var Q = require('q');
var PromisesSyncQueu = require('../psq');

var timestamp = (new Date()).getTime();
var arr = ['A', 'B', 'C'];

function doPromiseToItem(item){
  var deferred = Q.defer();
  var randomFactor = 10000 * Math.random()
  setTimeout(function(rand){
    console.log(item + ' solved after random ' + (Math.round((rand/1000)*100)/100) + ' secs.');
    deferred.resolve();
  }, randomFactor, randomFactor);
  return deferred.promise;
}

var promisesQueu = new PromisesSyncQueu();

for(var i=0; i<arr.length;i++){
  var item = arr[i];
  promisesQueu.addPromise(doPromiseToItem, item);
}

promisesQueu.resolveAllSync().then(function(){
  var duration = Math.round(((((new Date()).getTime()) - timestamp)/1000)*100)/100;
  console.log('ALL SOLVED in ' + duration + ' seconds.');
});

test/QAllTest.js
var Q = require('q');

var timestamp = (new Date()).getTime();
var arr = ['A', 'B', 'C'];

function doPromiseToItem(item){
  var deferred = Q.defer();
  var randomFactor = 10000 * Math.random()
  setTimeout(function(rand){
    console.log(item + ' solved after random ' + (Math.round((rand/1000)*100)/100) + ' secs.');
    deferred.resolve();
  }, randomFactor, randomFactor);
  return deferred.promise;
}

var promises = [];

for(var i=0; i<arr.length;i++){
  var item = arr[i];
  var promise = doPromiseToItem(item);
  promises.push(promise);
}

Q.all(promises).then(function(){
  var duration = Math.round(((((new Date()).getTime()) - timestamp)/1000)*100)/100;
  console.log('ALL SOLVED in ' + duration + ' seconds.');
});

test/PSQTestFail.js
var Q = require('q');
var PromisesSyncQueu = require('../psq');

var timestamp = (new Date()).getTime();
var arr = ['A', 'B', 'C', 'D'];

function doPromiseToItem(item){
  var deferred = Q.defer();
  var randomFactor = 10000 * Math.random()
  setTimeout(function(obj){
    if(obj.item === 'C'){
      console.log(item + ' NOT solved after random ' + (Math.round((obj.rand/1000)*100)/100) + ' secs.');
      deferred.reject(obj.item);
      return;
    }
    console.log(item + ' solved after random ' + (Math.round((obj.rand/1000)*100)/100) + ' secs.');
    deferred.resolve();
  }, randomFactor, {"rand":randomFactor,
                    "item" : item});
  return deferred.promise;
}

var promisesQueu = new PromisesSyncQueu();

for(var i=0; i<arr.length;i++){
  var item = arr[i];
  promisesQueu.addPromise(doPromiseToItem, item);
}

promisesQueu.resolveAllSync().fail(function(err){
  var duration = Math.round(((((new Date()).getTime()) - timestamp)/1000)*100)/100;
  console.log('FAILED because I found ' + err + ' after ' + duration + ' seconds.');
});

psq.js
var Q = require('q');

function Promise(fn, params){
  this.fn_ = fn;
  this.params_ = params;
}

function PromisesSyncQueu(){
  this.internalQueu_=[];
}

PromisesSyncQueu.prototype.addPromise = function(fn){
  var argsLength = arguments.length;
  var parameters = [];
  for(var i=1; i<argsLength; i++){
    parameters.push(arguments[i]);
  }
  var promise = new Promise(fn, parameters);
  this.internalQueu_.push(promise);
}

PromisesSyncQueu.prototype.resolveAllSync = function(){
  var deferred = Q.defer();
  var self = this;
  if(this.internalQueu_.length!==0){
   var promiseObj = this.internalQueu_.shift();
   promiseObj.fn_.apply(this, promiseObj.params_).then(function(){
        self.resolveAllSync().then(function(){
                                deferred.resolve();})
                             .fail(function(err){
                                deferred.reject(err);
                            });
    }).fail(function(err){
      deferred.reject(err);
    });
  }
  else{
    deferred.resolve();
  }
  return deferred.promise;
}

module.exports = PromisesSyncQueu;

psqueue.js
/*
  The MIT License (MIT)

  Copyright (c) 2015 Christian Adam

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
*/

/**
  @overview PSQueue module.
  @author Christian Adam
*/

/* jshint node:true */
var Q = require('q');

/**
 * A class representing a Promise
 * @param {function} fn A function returning a Q promise.
 * @param {object} params An array of params to pass to the function.
 */
function Promise(fn, params) {
  this.fn_ = fn;
  this.params_ = params;
}

/**
 * A class representing a Queue of promises
 * that will be solved synchronously.
 *
 */
function PSQueue() {
  this.internalQueu_ = [];
  this.resultsList = [];
}

/**
 * Adds a promise to the queue
 * @param {function} fn A function generating a promise.
 */
PSQueue.prototype.addPromise = function (fn) {
  if (typeof fn !== 'function') {
    throw new Error('You should add functions returning Q promises.');
  }
  var argsLength = arguments.length,
    parameters = [];
  for (var i = 1; i < argsLength; i++) {
    parameters.push(arguments[i]);
  }
  var promise = new Promise(fn, parameters);
  this.internalQueu_.push(promise);
};

/**
 * Resolves all the promises added to the queue in order.
 * @param   {object} torch An optional torch object that will be passed as last parameter in following promises.
 * @returns {object} A Q promise that will be solved once all promises are solved in order. The promise will resolve to an
 *                   array of values corresponding to the resolving values of each promise. If the promise fails it will
 *                   pass the corresponding error.
 */
PSQueue.prototype.resolveAllSync = function (torch) {
  var deferred = Q.defer(),
    self = this;
  if (this.internalQueu_.length !== 0) {
    var promiseObj = this.internalQueu_.shift();
    if (typeof torch === 'object') {
      promiseObj.params_.push(torch);
    }
    try {
      promiseObj.fn_.apply(this, promiseObj.params_).then(function (result) {
        self.resultsList.push(result);
        self.resolveAllSync(torch).then(function (resultsList) {
            deferred.resolve(resultsList);
          })
          .fail(function (err) {
            deferred.reject(err);
          });
      }).fail(function (err) {
        deferred.reject(err);
      });
    } catch (err) {
      deferred.reject(err);
    }
  } else {
    deferred.resolve(self.resultsList);
  }
  return deferred.promise;
};

/**
 * Alias for resolveAllSync
 */
PSQueue.prototype.allSync = function(torch){
  return this.resolveAllSync(torch);
};

/**
 * Alias for resolveAllSync
 */
PSQueue.prototype.all = function(torch){
  return this.resolveAllSync(torch);
};

module.exports = PSQueue;


demos/QAllTest.js
/*
  The MIT License (MIT)

  Copyright (c) 2015 Christian Adam

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
*/

/**
  @overview A test file for showing how Q.all works.
  @author Christian Adam
*/

/* jshint node:true */
var Q = require('q'),
  timestamp = (new Date()).getTime(),
  arr = ['A', 'B', 'C'],
  promises = [];

function doPromiseToItem(item) {
  var deferred = Q.defer();
  var randomFactor = 10000 * Math.random();
  setTimeout(function (rand) {
    console.log(item + ' solved after random ' + (Math.round((rand / 1000) * 100) / 100) + ' secs.');
    deferred.resolve(item);
  }, randomFactor, randomFactor);
  return deferred.promise;
}

for (var i = 0; i < arr.length; i++) {
  var item = arr[i];
  var promise = doPromiseToItem(item);
  promises.push(promise);
}

Q.all(promises).then(function (arrResults) {
  var duration = Math.round(((((new Date()).getTime()) - timestamp) / 1000) * 100) / 100;
  console.log('ALL SOLVED in ' + duration + ' seconds.');
  console.log('Results are:\n' + arrResults.join('\n'));
});


demos/PSQueueTest.js
/*
  The MIT License (MIT)

  Copyright (c) 2015 Christian Adam

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
*/

/**
  @overview A test file for showing how to use PSQueue. This test represent a happy scenario.
  @author Christian Adam
*/

/* jshint node:true */
var Q = require('q'),
  PSQueue = require('../psqueue'),
  timestamp = (new Date()).getTime(),
  arr = ['A', 'B', 'C'],
  promisesQueue = new PSQueue();

function doPromiseToItem(item, torch) {
  var deferred = Q.defer();
  var randomFactor = 10000 * Math.random();
  setTimeout(function (rand) {
    console.log(item + ' solved after ' + torch.text + ' taking random ' + (Math.round((rand / 1000) * 100) / 100) + ' secs.');
    torch.text = (torch.text === 'BEGGINING OF TIME' ? item : (torch.text + ',' + item));
    deferred.resolve(item + ':OK');
  }, randomFactor, randomFactor);
  return deferred.promise;
}

for (var i = 0; i < arr.length; i++) {
  var item = arr[i];
  promisesQueue.addPromise(doPromiseToItem, item);
}

promisesQueue.resolveAllSync({
  text: 'BEGGINING OF TIME'
}).then(function (arrResults) {
  var duration = Math.round(((((new Date()).getTime()) - timestamp) / 1000) * 100) / 100;
  console.log('ALL SOLVED in ' + duration + ' seconds.');
  console.log('Results are:\n' + arrResults.join('\n'));
});


demos/PSQueueTestFail.js
/*
  The MIT License (MIT)

  Copyright (c) 2015 Christian Adam

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in all
  copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
  SOFTWARE.
*/

/**
  @overview A test file for showing how to use PSQueue. This test represent a failing scenario.
  @author Christian Adam
*/

/* jshint node:true */
var Q = require('q'),
  PSQueue = require('../psqueue'),
  timestamp = (new Date()).getTime(),
  arr = ['A', 'B', 'C', 'D'],
  promisesQueue = new PSQueue();

function doPromiseToItem(item, torch) {
  var deferred = Q.defer();
  var randomFactor = 10000 * Math.random();
  setTimeout(function (obj) {
    if (obj.item === 'C') {
      console.log(item + ' NOT solved after random ' + (Math.round((obj.rand / 1000) * 100) / 100) + ' secs.');
      deferred.reject(obj.item);
      return;
    }
    console.log(item + ' solved after random ' + (Math.round((obj.rand / 1000) * 100) / 100) + ' secs.');
    torch.push(item);
    deferred.resolve();
  }, randomFactor, {
    "rand": randomFactor,
    "item": item
  });
  return deferred.promise;
}

for (var i = 0; i < arr.length; i++) {
  var item = arr[i];
  promisesQueue.addPromise(doPromiseToItem, item);
}

var partialResult = [];

promisesQueue.resolveAllSync(partialResult).fail(function (err, arrResults) {
  var duration = Math.round(((((new Date()).getTime()) - timestamp) / 1000) * 100) / 100;
  console.log('FAILED because I found ' + err + ' after ' + duration + ' seconds.');
  console.log('Partial result is: ' + partialResult.join(','));
});
